### PR TITLE
Parenthesize interpolations containing `global::`

### LINKED ISSUE #3447: `global::` in string interpolation
### Input code

```cs
return $"<b>{(global::ParadoxNotion.ReflectionTools.FriendlyName(varType))}</b>";
```

### Erroneous output

```c#
return $"<b>{global::ParadoxNotion.ReflectionTools.FriendlyName(varType)}</b>";
```

The output fails to re-compile and produces the following compiler error message:

```
CS0103: The name 'global' does not exist in the current context
```

The parentheses are necessary because otherwise `:` specifies a numeric format.

### Details
* Product in use: ICSharpCode.Decompiler nuget package
* Version in use: fd6070d2292edcf817763b191ea4d491368970ec
* As always, I am willing to contribute a fix. However, I may need advice on the best way to solve this issue.

### PR DESCRIPTION
**Pull request by ds5678**
Resolves #3447 

### Solution
* Any comments on the approach taken, its consistency with surrounding code, etc.
  * As recommended, I implemented the fix in the `InsertParenthesesVisitor` class.
* Which part of this PR is most in need of attention/improvement?
  * I used recursion to check if any child has `global::`. Is there a better (more performant) way to check?
* [x] At least one test covering the code changed

### CODE REVIEW

**Comment by siegfriedpammer** [comment]
Ad recursion:

You can avoid stack overflows by using `AstNode.Dsecendants`, see https://github.com/icsharpcode/ILSpy/blob/03b7444943e720b3134d296c0c8dd3876f8ea4ce/ICSharpCode.Decompiler/CSharp/Syntax/AstNode.cs#L291-L337

**Comment by siegfriedpammer** [comment]
Note to self: do not review pull requests when not entirely awake :(

sorry, I gave wrong feedback. We cannot prevent stack overflows in this case, because the whole thing is using the Visitor pattern. So the problem remains and the performance might even be worse due to the use of LINQ.

So, is there actually anything we could do to limit the number of nodes to check? Can we, for example, stop at anonymous functions/lambdas? Do we still need to disambiguate inside invocation argument lists? Might we be able to just look at the top-level expression?

This is fine according to [SharpLab](https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQMwAJboMLoN7LpGYZRwAM6AsgBRmUCGAlPoce1AOzoAkARABUApgGcALiHy0A5gBsA9gCMGskCDIA6AKIA7AG4BLAE7ydAWyE6xG7AFcjRy2IAixoQGMx8owE90YdGYAXz4AbjYiIIiSTApqOgBWAB56ABp0RXl5WQA+dAAPFgIkdg5ufmFxSTxgIVkhaQYxITo4nyLMbjklFTVNXUMTcycbe0crV0dPbx8NABlLaTEAC3Q88lD0IJDwkuIopCCgA=)...

```
    public string M(string a) {
        return $"Test: {M(global::System.Environment.CurrentDirectory + a)}";
    }
    public string M() {
        return $"Test: {delegate(string y) { return global::System.Environment.CurrentDirectory.Length > 0; }}";
    }
```

So I guess, we can stop descending as soon as we find:
* Parentheses
* An argument
* An anonymous method/lambda expression

What do you think?

**Comment by ds5678** [comment]
I implemented what you suggested. However, I encountered another issue that's preventing my tests from passing.

**Comment by siegfriedpammer** [comment]
While fixing #3464 I noticed that `public static string Lambda1 => $"Prefix {() => global::System.DateTime.Now} suffix";` doesn't require a cast, but still requires parentheses.

File: ICSharpCode.Decompiler/CSharp/OutputVisitor/InsertParenthesesVisitor.cs [hunk 1]
```diff
@@ -389,6 +390,36 @@ public override void VisitAsExpression(AsExpression asExpression)
...
+			base.VisitInterpolation(interpolation);
+
+			static bool InterpolationNeedsParenthesized(AstNode node)
+			{
+				if (node is MemberType { IsDoubleColon: true })
+					return true;
+
+				if (node is ConditionalExpression)
+					return false; // Conditional expressions are parenthesized in their own visit method.
+				if (node is AnonymousMethodExpression or LambdaExpression)
```
Comment (siegfriedpammer) [inline ICSharpCode.Decompiler/CSharp/OutputVisitor/InsertParenthesesVisitor.cs:407]:
```suggestion
				if (node is AnonymousMethodExpression or LambdaExpression { Body: BlockStatement })
```

File: ICSharpCode.Decompiler/CSharp/OutputVisitor/InsertParenthesesVisitor.cs [hunk 2]
```diff
@@ -389,6 +390,36 @@ public override void VisitAsExpression(AsExpression asExpression)
...
 		}
 
+		public override void VisitInterpolation(Interpolation interpolation)
+		{
+			// If an interpolation contains global::, we need to parenthesize the expression.
+			if (InterpolationNeedsParenthesized(interpolation))
+				Parenthesize(interpolation.Expression);
+			base.VisitInterpolation(interpolation);
+
+			static bool InterpolationNeedsParenthesized(AstNode node)
```
Comment (siegfriedpammer) [inline ICSharpCode.Decompiler/CSharp/OutputVisitor/InsertParenthesesVisitor.cs:400]:
```suggestion
			static bool InterpolationNeedsParenthesis(AstNode node)
```

File: ICSharpCode.Decompiler/CSharp/OutputVisitor/InsertParenthesesVisitor.cs [hunk 3]
```diff
@@ -389,6 +390,36 @@ public override void VisitAsExpression(AsExpression asExpression)
...
+					return true;
+
+				if (node is ConditionalExpression)
+					return false; // Conditional expressions are parenthesized in their own visit method.
+				if (node is AnonymousMethodExpression or LambdaExpression)
+					return false;
+				if (node is InvocationExpression invocation)
+					return InterpolationNeedsParenthesized(invocation.Target);
+				if (node is CastExpression cast)
+					return InterpolationNeedsParenthesized(cast.Expression);
```
Comment (siegfriedpammer) [inline ICSharpCode.Decompiler/CSharp/OutputVisitor/InsertParenthesesVisitor.cs:412]: There is a whole lot of special cases for cast-expressions: see `InsertParenthesesVisitor.VisitCastExpression`.

Comment (ds5678) [reply ICSharpCode.Decompiler/CSharp/OutputVisitor/InsertParenthesesVisitor.cs]: I reordered the visitation of children, so that those special cases should get hit before we reason about whether or not to parenthesize the interpolation expression.

**Comment by ds5678** [comment]
> While fixing #3464 I noticed that `public static string Lambda1 => $"Prefix {() => global::System.DateTime.Now} suffix";` doesn't require a cast, but still requires parentheses.

This is a recent change caused by you fixing #3464. Previously, the decompiler would emit the cast always. I added a pragma directive in the unit test to choose the appropriate format based on framework and C# version.

File: ICSharpCode.Decompiler.Tests/TestCases/Pretty/GloballyQualifiedTypeInStringInterpolation.cs [hunk 4]
```diff
@@ -0,0 +1,25 @@
+﻿namespace ICSharpCode.Decompiler.Tests.TestCases.Pretty
+{
+	public static class GloballyQualifiedTypeInStringInterpolation
+	{
+		public static string Root => $"Prefix {(global::System.DateTime.Now)} suffix";
+		public static string Cast => $"Prefix {((int)global::System.DateTime.Now.Ticks)} suffix";
+#if CS100 && !NET40
```
Comment (ds5678) [inline ICSharpCode.Decompiler.Tests/TestCases/Pretty/GloballyQualifiedTypeInStringInterpolation.cs:7]: Should this have been `#if CS100 && NET60` instead?

Comment (ds5678) [reply ICSharpCode.Decompiler.Tests/TestCases/Pretty/GloballyQualifiedTypeInStringInterpolation.cs]: I think so.

**Comment by ds5678** [comment]
Also, I notice that you seem to normal merge all pull requests, even though mine almost always have a few "junk" commits. Is there a reason you don't squash merge?

File: ICSharpCode.Decompiler.Tests/TestCases/Pretty/GloballyQualifiedTypeInStringInterpolation.cs (hunk 4 above)
Comment (siegfriedpammer) [reply ICSharpCode.Decompiler.Tests/TestCases/Pretty/GloballyQualifiedTypeInStringInterpolation.cs]: I think it makes more sense to use `!NET40` because classic framework is the "legacy edge case" and we want to minimize necessary changes when updating the framework version.

Comment (ds5678) [reply ICSharpCode.Decompiler.Tests/TestCases/Pretty/GloballyQualifiedTypeInStringInterpolation.cs]: This happens because `DefaultInterpolatedStringHandler` is only available on .NET 6 and above. If any of our tests ran on .NET 5, they would fail.

Comment (ds5678) [reply ICSharpCode.Decompiler.Tests/TestCases/Pretty/GloballyQualifiedTypeInStringInterpolation.cs]: Do you want me to change it back?

Comment (siegfriedpammer) [reply ICSharpCode.Decompiler.Tests/TestCases/Pretty/GloballyQualifiedTypeInStringInterpolation.cs]: Hmmm... yes, and because we emit conditional compilation symbols for all subsumed versions used, it works. OK, we can leave it like this...

**Comment by siegfriedpammer** [comment]
> This is a recent change caused by you fixing #3464.

Yes, my comment was imprecise: While playing around with your test case (in order to decide whether I can pass `allowImplicitConversion: true` to `ConvertTo`), I noticed that Visual Studio highlights the cast to `Func<...>` as unnecessary, so I tried removing it and it seemed to work, but the parenthesis was no longer optional, if the body of the lambda was just an expression containing `global::`.

> Also, I notice that you seem to normal merge all pull requests, even though mine almost always have a few "junk" commits. Is there a reason you don't squash merge?

I usually organize my commits into smaller self-contained units/changes to make bisecting after bug reports easier. If a PR has somewhat organized commits I usually just do a normal merge, however, if there is junk in there, I do squash-merges too. It all the depends on the person creating the PR.

**Comment by ds5678** [comment]
Is there something else I need to do for this pull request to be merged?

File: ICSharpCode.Decompiler/CSharp/OutputVisitor/InsertParenthesesVisitor.cs [hunk 5]
```diff
@@ -17,6 +17,7 @@
 // DEALINGS IN THE SOFTWARE.
 
 using System;
+using System.Linq;
```
Comment (siegfriedpammer) [inline ICSharpCode.Decompiler/CSharp/OutputVisitor/InsertParenthesesVisitor.cs:20]: Nit: no longer necessary? 

```suggestion
```

**Comment by siegfriedpammer** [comment]
> Is there something else I need to do for this pull request to be merged?

No, I just wasn't sure whether you are done yet.

**Comment by siegfriedpammer** [comment]
Thank you very much for your continued contributions!

**Comment by ds5678** [comment]
No problem. I appreciate your reviews.
